internal/cli/cmd/args.go

```diff
@@ -92,46 +89,53 @@ type RegistryMatcher interface {
-func NewNamespaceMatcherFunc(matcher NamespaceMatcher) ValidArgsFunc {
+func NewEpinioArgValidator(matcher *EpinioMatcher) ValidArgsFunc {
 	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
-		if len(args) != 0 {
+		if matcher.ArgLimit != -1 && len(args) >= matcher.ArgLimit {
 			return nil, cobra.ShellCompDirectiveNoFileComp
 		}
 
-		matcher.GetAPI().DisableVersionWarning()
+		// the len of the args indicates the matcher to apply
+		// i.e.: when we are looking for the first arg len(arg) is 0, then the matcher to apply is the first
+		matcherIndex := len(args)
 
-		matches := matcher.NamespacesMatching(toComplete)
-		return matches, cobra.ShellCompDirectiveNoFileComp
-	}
-}
+		// If the ArgLimit is negative we need to "loop" over the matchers.
+		// This is used to get the "any" validator
+		if matcher.ArgLimit < 0 {
+			matcherIndex = len(args) % len(matcher.Matchers)
+		}
 
-// NewConfigurationMatcherFirstFunc returns a function returning list of matching configurations from the
-// provided partial command.  It only matches for the first command argument.
-func NewConfigurationMatcherFirstFunc(matcher ConfigurationMatcher) ValidArgsFunc {
-	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
-		if len(args) != 0 {
+		// this shoud never happen, but better to avoid a nil pointer
+		if len(matcher.Matchers)-1 < matcherIndex {
 			return nil, cobra.ShellCompDirectiveNoFileComp
 		}
 
-		matcher.GetAPI().DisableVersionWarning()
+		matcherFn := matcher.Matchers[matcherIndex]
+		matches := matcherFn(toComplete)
+
+		// apply filter for already matched resources
+		if matcher.Filter != nil {
+			matches = matcher.Filter(args, matches)
+		}
 
-		matches := matcher.ConfigurationMatching(toComplete)
 		return matches, cobra.ShellCompDirectiveNoFileComp
 	}
 }
 
-// NewConfigurationMatcherAnyFunc returns a function returning a list of matching configurations
-// from the provided partial command.  It matches for all command arguments
-func NewConfigurationMatcherAnyFunc(matcher ConfigurationMatcher) ValidArgsFunc {
-	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
-		matcher.GetAPI().DisableVersionWarning()
+// NewNamespaceMatcherFunc returns a function returning list of matching namespaces from the
+// provided partial command.  It only matches for the first command argument.
 
-		matches := FilteredMatchingFinder(args, toComplete, matcher.ConfigurationMatching)
+func FirstArgValidator(matcher MatcherFunc) ValidArgsFunc {
+	return NewEpinioArgValidator(NewEpinioMatcher(matcher))
+}
 
-		return matches, cobra.ShellCompDirectiveNoFileComp
-	}
+func AnyArgsValidator(matcher MatcherFunc) ValidArgsFunc {
+	epinioMatcher := NewEpinioMatcher(matcher)
+
+	epinioMatcher.Filter = FilterMatches
+	epinioMatcher.ArgLimit = -1
```

Comment:
For future PRs extending the set of matchers built on this framework, maybe add `Any()` and `Filter(filter)` modifiers which take an `EpinioMatcher`, make the change, and return the changed struct. Then the next line can be written as
```
return NewEpinioArgValidator(NewEpinioMatcher(matcher).Any().Filter (FilterMatches))
```

Reply:
Yes, makes sense! Thanks